src/move/research/prover-examples/sources/value_holder.move

```diff
@@ -0,0 +1,90 @@
+module prover_examples::value_holder {
+    use std::signer;
+
+    const DEFAULT_VALUE: u8 = 0;
+
+    struct ValueHolder has key {
+        value: u8
+    }
+
+    struct ValueHolderManifest has key {
+        /// The addresses of all accounts that have a ValueHolder.
+        account_addresses: vector<address>,
+        /// The value that all ValueHolders should have.
+        value: u8
+    }
+
+    spec schema Initialized {
+        aborts_if !exists<ValueHolderManifest>(@prover_examples);
+    }
+
+    spec module {
+        apply Initialized to * except init_module;
+    }
+
+    invariant update[suspendable] forall account_address: address where !old(
+        exists<ValueHolder>(account_address)
+    ) && exists<ValueHolder>(account_address):
+        contains(
+            global<ValueHolderManifest>(@prover_examples).account_addresses,
+            account_address
+        ) && old(
+            !contains(
+                global<ValueHolderManifest>(@prover_examples).account_addresses,
+                account_address
+            )
+        );
+
+    public fun init_value_holder(account: &signer) acquires ValueHolderManifest {
+        let value_holder_manifest_ref_mut = &mut ValueHolderManifest[@prover_examples];
+        move_to(
+            account,
+            ValueHolder { value: value_holder_manifest_ref_mut.value }
+        );
+        value_holder_manifest_ref_mut.account_addresses.push_back(
+            signer::address_of(account)
+        );
+    }
+
+    fun init_module(prover_examples: &signer) {
+        move_to(
+            prover_examples,
+            ValueHolderManifest { account_addresses: vector[], value: DEFAULT_VALUE }
+        );
+    }
+
+    spec init_module {
+        requires signer::address_of(prover_examples) == @prover_examples;
+        ensures old(!exists<ValueHolderManifest>(@prover_examples));
+        ensures exists<ValueHolderManifest>(@prover_examples);
+        ensures global<ValueHolderManifest>(signer::address_of(prover_examples))
+            == ValueHolderManifest { account_addresses: vector[], value: DEFAULT_VALUE };
+    }
+
+    spec init_value_holder {
+        pragma disable_invariants_in_body;
+
+        let account_address = signer::address_of(account);
+        requires !contains(
+            global<ValueHolderManifest>(@prover_examples).account_addresses,
+            account_address
+        );
```

Comment:
@rahxephon89  I'm noticing in [the `requires` docs](https://aptos.dev/en/build/smart-contracts/prover/spec-lang#requires-condition) the following:

> Specifically, `requires` should be avoided for public APIs

I've been able to eliminate a requirement that the manifest exists via `schema Initialized`, but is there  a best practice for eliminating the highlighted `requires` condition? I'm confused that I even need it, because right now there is now way to get an address into the manifest *without* using this function - so why would the prover fail without the highlighted `requires`?

I've found that I can replace the first one with the following:

```

Reply:
Translating `requires` into a corresponding `aborts_if` condition would be ideal for public APIs. In this example, if there is no `requires` in `init_value_holder`, the clause `old(
            !contains(
                global<ValueHolderManifest>(@prover_examples).account_addresses,
                account_address
            )` in the global invariant will not hold

Reply:
@rahxephon89 thanks for the explanation. I decided to use more expressive invariants in a4ee3b8, which allowed me to check that functions which will violate the pre-condition are flagged as such - for example the one removed in c5cefc8